### PR TITLE
`Iris`: Add Iris chat in text exercises

### PR DESCRIPTION
**Pull request by MichaelOwenDyer**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [ ] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Up until now, Iris has only been available for programming exercises. We want to expand the availability of Iris to other exercise types.

### Description

Iris has been added to text exercises.
- A new Session type has been added and corresponding SessionService.
- The Iris button now appears in text exercises when the feature is enabled globally, in the course, and in the exercise itself. Clicking the button opens the Iris chat window.
- A [new Pyris pipeline](https://github.com/ls1intum/Pyris/pull/161) has been added to power the feature.

### Future Work

In this PR, Iris is only aware of the last *submission* to the exercise. If the user edits the text without submitting, Iris will not see it. This is because it is not possible with the current server architecture for the client to send form data for use in the Pyris feature pipeline. This is outside the scope of this PR.

### Steps for Testing

Prerequisites:
- Iris is enabled globally
- 1 Course
- 1 Text exercise

1. Log in to Artemis
2. Navigate to Course Administration, click on your course
3. Navigate to Iris settings and enable Iris Text Exercise Chat
4. Navigate to the settings of text exercise in the course and enable Text Exercise Chat
5. Navigate to the exercise as a student
6. See that the Iris button appears and opens the chat window
7. Talk to Iris, ask for help solving the exercise
8. Submit some solution
9. Continue talking to Iris

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review

- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance even for very large courses with more than 2000 students.
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance even for very large courses with more than 2000 students.
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2
#### Exam Mode Test
- [ ] Test 1
- [ ] Test 2
#### Performance Tests
- [ ] Test 1
- [ ] Test 2

## Summary by CodeRabbit

- **New Features**
	- Introduced `IrisTextExerciseChatSession` for managing chat sessions related to text exercises.
	- Added support for dynamic variant handling in chat settings based on course configurations.
	- Added new DTOs for managing text exercise chat status updates and pipeline executions.

- **Bug Fixes**
	- Improved handling of rate limits and access checks for chat sessions.

- **Documentation**
	- Enhanced guidelines for database relationships and fetching strategies, emphasizing best practices for lazy fetching.

- **Chores**
	- Removed deprecated classes and methods related to the Hestia integration.
	- Updated repository interfaces to reflect new chat session functionalities.
	- Adjusted coverage thresholds in testing configuration.

### CODE REVIEW

File: src/main/webapp/i18n/de/iris.json [hunk 1]
```diff
@@ -62,7 +67,7 @@
                     "breadcrumb": "Iris Einstellungen",
                     "global": "Globale Iris Einstellungen",
                     "course": "Kurs Iris Einstellungen",
-                    "programmingExercise": "Programmieraufgabe Iris Einstellungen"
+                    "exercise": "Programmieraufgabe Iris Einstellungen"
```
Comment (kaancayli) [inline src/main/webapp/i18n/de/iris.json:70]:
```suggestion
                    "exercise": "Aufgabe Iris Einstellungen"
```

File: src/test/java/de/tum/cit/aet/artemis/iris/AbstractIrisIntegrationTest.java [hunk 2]
```diff
@@ -58,9 +58,13 @@ void tearDown() throws Exception {
 
     protected void activateIrisGlobally() {
         var globalSettings = irisSettingsService.getGlobalSettings();
+        // if (globalSettings.getIrisTextExerciseChatSettings() == null) {
+        // globalSettings.setIrisTextExerciseChatSettings(new IrisTextExerciseChatSubSettings());
+        // }
```
Comment (kaancayli) [inline src/test/java/de/tum/cit/aet/artemis/iris/AbstractIrisIntegrationTest.java:63]: I would remove this, if it's not used

**Review by kaancayli — CHANGES_REQUESTED** [review]
Overall good code. Couple of suggestions

File: src/main/webapp/i18n/de/iris.json (hunk 1 above)
Comment (MichaelOwenDyer) [reply src/main/webapp/i18n/de/iris.json:70]: This is actually a temporary fix to make the language files consistent. The client code expects this translation to be under "exercise". But it should be corrected in a later PR with all the rest of the renaming of the programming exercise chat feature (currently just called "exercise chat" across the codebase)

**Review by yassinsws — APPROVED** [review]
Coder looks good

**Review by kaancayli — APPROVED** [review]
LGTM

**Pull request merged**